pkg/rabbit/connections_handler.go

```diff
@@ -119,23 +118,35 @@ func (r *RabbitMQHelper) SetupRabbitMQConnectionAndChannel(
 	configFunction func(RabbitMQConnectionInterface, RabbitMQChannelInterface) error,
 	dialFunc func(string) (RabbitMQConnectionWrapperInterface, error)) {
 	var err error
-	retryConnection, retryChannel := true, true
-	for retryConnection || retryChannel {
+	for {
 		// Wait one cycle duration always except the first time
 		if !r.firstSetup {
 			time.Sleep(time.Millisecond * r.cycleDuration)
 		} else {
 			r.logger.Info("Creating and Configuring RabbitMQ Connection and Channel")
 			r.firstSetup = false
 		}
-		r.Connection, r.Channel, err = r.CreateAndConfigConnectionsAndChannel(&retryConnection, &retryChannel, rabbitMQURL, configFunction, dialFunc)
-		if err != nil {
-			r.logger.Error(err)
-			if retryConnection {
-				r.CloseRabbitMQConnections()
-			}
+		// create the connection if needed
+		if r.Connection == nil || r.Connection.IsClosed() {
+			r.Channel = nil
+			r.Connection, err = r.CreateConnection(rabbitMQURL, dialFunc)
+		}
+		// create the channel if needed
+		if err == nil && (r.Channel == nil || r.Channel.IsClosed()) {
+			r.Channel, err = r.CreateChannel()
+		}
+		// config channel and connection
+		if err == nil && configFunction != nil {
+			err = r.ConfigConnectionAndChannel(configFunction)
+		}
+		// setup completed successfully
+		if err == nil {
+			break
 		}
+		r.logger.Error(err)
+		r.CloseRabbitMQConnections()
 	}
+	// watch for any connection unexpected closures
 	if r.firstSetup {
 		go r.WatchRabbitMQConnections(ctx, rabbitMQURL, configFunction, dialFunc)
```

Comment:
This got me thinking about the `WatchRabbitMQConnections` function. The `firstSetup` is what prevents the watch function from getting called multiple times should `SetupRabbitMQConnectionAndChannel` be called multiple times. `WatchRabbitMQConnections` is in itself public as well so a user can just directly call it multiple times. The function will run indefinitely until the context is cancelled so we definitely don't want more than 1. 

It seems we should have a lock to guarantee only one instance is watching for connection/channel closures.